geode-core/src/distributedTest/java/org/apache/geode/internal/cache/execute/ClientServerFunctionExecutionDUnitTest.java

```diff
@@ -424,13 +437,73 @@ public void testOnServerExecution_FunctionInvocationTargetException() {
             Boolean.FALSE));
   }
 
+  @Test
+  public void testNestedFunctionExecutionOnReplicatedRegion() {
+    // Set function execution timeout in the client
+    client.invoke(() -> setFunctionExecutionTimeout(30000));
+
+    // Create single client/server scenario with replicated region
+    createSingleClientServerScenario();
+
+    // Execute the function with multiple threads
+    function = new TestFunction(true, TEST_EXECUTE_PARENT);
+    client.invoke(() -> executeFunction(function, MAX_FE_THREADS * 3));
+  }
+
+  private void setFunctionExecutionTimeout(int timeout) {
+    System.setProperty("gemfire.CLIENT_FUNCTION_TIMEOUT", String.valueOf(timeout));
+  }
 
   private void createScenario() {
     LogWriterUtils.getLogWriter()
         .info("ClientServerFFunctionExecutionDUnitTest#createScenario : creating scenario");
     createClientServerScenarionWithoutRegion();
   }
 
+  private void createSingleClientServerScenario() {
+    createCacheInSingleClientServer();
+    Integer port = server1.invoke(() -> createCacheServer());
+    serverPort1 = port;
+    server1.invoke(() -> createReplicatedRegion());
+    client.invoke(() -> createPool(NetworkUtils.getServerHostName(), port));
+    client.invoke(() -> createProxyRegion(null));
+  }
+
+  private void createPool(String host, Integer port) {
+    pool = (PoolImpl) PoolManager.createFactory().addServer(host, port).setRetryAttempts(0)
+        .create("PRClientServerTestBase");
+  }
+
+  private void executeFunction(Function function, int numThreads) {
+    final AtomicInteger successfulFunctionExecutions = new AtomicInteger(0);
+    final AtomicInteger failedFunctionExecutions = new AtomicInteger(0);
+
+    // Execute the function with multiple threads
+    Thread[] threads = new Thread[numThreads];
```

Comment:
Functionally this might not be a big deal, but there is a new ConcurrencyRule that offers helper methods to add several runnables or callables, then execute them in parallel.  It might save you a few lines of code here and it uses an ExecutorService under the hood.  You should be able to search for ConcurrencyRule and see usages in our code.

Reply:
Thanks! I'll look into that rule.

Reply:
Another option is the ExecutorServiceRule in geode-junit. If you're firing off threads one at a time and don't need to line them up at a start line to start them all at once then ExecutorServiceRule is probably the better option.